crates/iceberg/src/puffin/metadata.rs

```diff
@@ -0,0 +1,809 @@
+    pub(crate) properties: HashMap<String, String>,
+}
+
+impl FileMetadata {
+    pub(crate) const MAGIC_LENGTH: u8 = 4;
+    pub(crate) const MAGIC: [u8; FileMetadata::MAGIC_LENGTH as usize] = [0x50, 0x46, 0x41, 0x31];
+
+    // We use the term FOOTER_STRUCT to refer to the fixed-length portion of the Footer, as illustrated below.
+    //
+    //                        Footer
+    //                          |
+    //  -------------------------------------------------
+    // |                                                 |
+    // Magic FooterPayload FooterPayloadLength Flags Magic
+    //                     |                             |
+    //                      -----------------------------
+    //                                    |
+    //                              FOOTER_STRUCT
+
+    const FOOTER_STRUCT_PAYLOAD_LENGTH_OFFSET: u8 = 0;
+    const FOOTER_STRUCT_PAYLOAD_LENGTH_LENGTH: u8 = 4;
+    const FOOTER_STRUCT_FLAGS_OFFSET: u8 = FileMetadata::FOOTER_STRUCT_PAYLOAD_LENGTH_OFFSET
+        + FileMetadata::FOOTER_STRUCT_PAYLOAD_LENGTH_LENGTH;
+    pub(crate) const FOOTER_STRUCT_FLAGS_LENGTH: u8 = 4;
+    const FOOTER_STRUCT_MAGIC_OFFSET: u8 =
+        FileMetadata::FOOTER_STRUCT_FLAGS_OFFSET + FileMetadata::FOOTER_STRUCT_FLAGS_LENGTH;
+    pub(crate) const FOOTER_STRUCT_LENGTH: u8 =
+        FileMetadata::FOOTER_STRUCT_MAGIC_OFFSET + FileMetadata::MAGIC_LENGTH;
+
+    fn check_magic(bytes: &[u8]) -> Result<()> {
+        if bytes != FileMetadata::MAGIC {
+            Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Bad magic value: {:?} should be {:?}",
+                    bytes,
+                    FileMetadata::MAGIC
+                ),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    async fn read_footer_payload_length(
+        file_read: &dyn FileRead,
+        input_file_length: u64,
+    ) -> Result<u32> {
+        let start = input_file_length - u64::from(FileMetadata::FOOTER_STRUCT_LENGTH);
+        let end = start + u64::from(FileMetadata::FOOTER_STRUCT_PAYLOAD_LENGTH_LENGTH);
+        let footer_payload_length_bytes = file_read.read(start..end).await?;
+        let mut buf = [0; 4];
+        buf.copy_from_slice(&footer_payload_length_bytes);
+        let footer_payload_length = u32::from_le_bytes(buf);
+        Ok(footer_payload_length)
+    }
+
+    async fn read_footer_bytes(
+        file_read: &dyn FileRead,
+        input_file_length: u64,
+        footer_payload_length: u32,
+    ) -> Result<Bytes> {
+        let footer_length = u64::from(footer_payload_length)
+            + u64::from(FileMetadata::FOOTER_STRUCT_LENGTH)
+            + u64::from(FileMetadata::MAGIC_LENGTH);
+        let start = input_file_length - footer_length;
+        let end = input_file_length;
+        file_read.read(start..end).await
+    }
+
+    fn err_out_of_bounds<T>() -> Result<T> {
```

Comment:
If this error doesn't occur often, how about writing to them directly? I believe we won't need to change them frequently. I tend to optimize for reading code.